### PR TITLE
History: update document URL on pushState/replaceState

### LINKED ISSUE #19157: Update document URL when a new history state is pushed/replaced
Part of #10992

https://html.spec.whatwg.org/multipage/#dom-history-pushstate Step 10

### PR DESCRIPTION
**Pull request by talklittle**
Update Document URL when calling pushState and replaceState on History.

https://html.spec.whatwg.org/multipage/#dom-history-pushstate
Steps 6, 7, 10

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #19157 (github issue number if applicable).

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/20453)

### CODE REVIEW

**Comment by jdm** [comment]
cc @cbrewster

File: components/script/dom/history.rs [hunk 1]
```diff
@@ -86,17 +87,56 @@ impl History {
...
+                // Step 6.5
+                if new_url.origin() != document_url.origin() {
+                    return Err(Error::Security);
+                }
+
+                new_url
+            },
+            // Step 7
+            None => {
+                let browsing_context_id = self.window.window_proxy().browsing_context_id();
```
Comment (cbrewster) [inline components/script/dom/history.rs]: I think we should be able to get the URL from the document itself rather than going via the constellation.

**Review by cbrewster — CHANGES_REQUESTED** [review]
@talklittle sorry for taking a while on this one!

Looks good, I think we can get the current entry URL from the document rather than going to the constellation.

This will need to be rebased since #20638. The next steps will be to make the constellation aware of the URL in the session history. If you'd like to do that work as a followup, I'd be happy to mentor. (ping me on mozilla IRC: cbrewster)

**Comment by talklittle** [comment]
@cbrewster Thanks for the review. I've rebased and applied your suggestions.

> The next steps will be to make the constellation aware of the URL in the session history. If you'd like to do that work as a followup, I'd be happy to mentor.

Thanks for the offer, but I am becoming busy again at the moment. If you want to continue the rest of the History work please feel free.

**Comment by cbrewster** [comment]
@talklittle no worries, thanks for doing this!

@bors-servo r+

**Comment by cbrewster** [comment]
```
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "location search", "test": "/html/browsers/history/the-location-interface/location_search.html", "line": 66417, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "Check if button.formAction is the document's new address when formaction content attribute is missing and pushState has been used", "test": "/html/semantics/forms/attributes-common-to-form-controls/formAction_document_address.html", "line": 113350, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "Check if input.formAction is the document's new address when formaction content attribute is missing and pushState has been used", "test": "/html/semantics/forms/attributes-common-to-form-controls/formAction_document_address.html", "line": 113351, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "location hash", "test": "/html/browsers/history/the-location-interface/location_hash.html", "line": 129937, "action": "test_result", "expected": "FAIL"}
```

**Comment by asajeffrey** [comment]
Yay! Lots more passing tests! @bors-servo r=cbrewster

**Comment by asajeffrey** [comment]
@bors-servo retry
 * infra

**Comment by cbrewster** [comment]
@bors-servo retry
 * servo-mac7 is broken, hopefully its not used this time

**Pull request merged**